---
title: " Study II - Importing HumaNn profiles into phyloseq"
author: "Florentin Constancias"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document: 
    keep_md: no
    toc: yes
    toc_float: true
    theme: flatly # Optional: Bootstrap theme
    highlight: tango # Syntax highlighting
    css: styles.css # Custom CSS for styling
---

```{r setup, include=FALSE}
rm(list = ls())
gc()

invisible(lapply(paste0('package:', names(sessionInfo()$otherPkgs)), detach, character.only=TRUE, unload=TRUE))
# options(java.parameters = "-Xmx80000m")
```


```{r setup2, include=FALSE}
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
##knitr::opts_chunk$set(cache=TRUE)
# knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(dev = "ragg_png") # https://ragg.r-lib.org/index.html
##knitr::opts_chunk$set(fig.width = 6) #https://r4ds.had.co.nz/graphics-for-communication.html#figure-sizing
knitr::opts_chunk$set(fig.asp = 0.618)
knitr::opts_chunk$set(fig.show = "hold")
knitr::opts_chunk$set(fig.show = "70%")
knitr::opts_chunk$set(fig.align = "center")
```

```{r setup3, include=FALSE}
library(tidyverse)
library(readxl)
library(readr)
library(speedyseq)

source_dir = "https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/"
# source_dir = "~/Documents/GitHub/DivComAnalyses/R/"

source(paste0(source_dir,"phyloseq_biobakery_functions.R"))
source(paste0(source_dir,"phyloseq_varia.R"))
source(paste0(source_dir,"phyloseq_alpha.R"))
source(paste0(source_dir,"phyloseq_beta.R"))
source(paste0(source_dir,"phyloseq_normalisation.R"))


source("https://raw.githubusercontent.com/fconstancias/metabaRpipe-source/master/Rscripts/functions.R")
source("https://raw.githubusercontent.com/fconstancias/KU_Caroline/refs/heads/main/code/functions/phyloseq_functions.R")

```


# Import and get ready:

Import previous metaphlan data so we will get the metadata from there:

```{r}
load(here::here("../../data/processed_data/metaphlan/01_data.Rdata"))
```

Define path of the PWY and Gene family table.- output of HumAn

```{r}
pathway <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_pathabundance_cpm.tsv"
kegg <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_genefamilies_uniref90_ko_renamed_relab.tsv"
l4c <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_genefamilies_uniref90_level4ec_renamed_relab.tsv"
go <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_genefamilies_uniref90_go_renamed_relab.tsv"
eggnog <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_genefamilies_uniref90_eggnog_relab.tsv"
rxn <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_genefamilies_uniref90_rxn_renamed_relab.tsv"
pfam <- "/Users/ljc444/Documents/Caroline/temp_results/humann/Merged_genefamilies_uniref90_pfam_renamed_relab.tsv"
```


```{r, eval=TRUE}
#' Import and Process HUMAnN Pathway Data for Microbial Ecology Analysis
#'
#' This function imports HUMAnN pathway data, processes it into a `phyloseq` object, 
#' and cleans and annotates the data to facilitate downstream analysis in microbial ecology studies.
#' 
#' @param pathway A file path or object containing HUMAnN pathway or gene family level data.
#' @param rank The taxonomic rank used for filtering or organizing the taxonomy table. Default is `"pathway"`.
#' @param tax_clean_gsub A regular expression pattern to clean unwanted characters or prefixes in taxonomy labels. Default is `"[a-t]__"`.
#' @param no_sel A vector of values to exclude during filtering, such as `NA` or `"unclassified"`. Default is `c(NA, "<NA>", "NA", "unclassified")`.
#' @param sub_pat_a A string pattern for initial cleaning of sample names. Default is `"_trimmed_HF_merged_Abundance-RELAB"`.
#' @param sub_pat_b A string pattern for additional cleaning of sample names. Default is `"_subset50"`.
#' @param tax_na_if A string indicating taxonomy entries to replace with `NA`. Default is `"unclassified"`.
#' @param sample_name_clean_a A regular expression pattern for extracting meaningful parts of sample names. Default is `"[^_]+"`.
#' @param sample_name_clean_b A regular expression for removing specific patterns in sample names. Default is `"[A-Z]"`.
#' @param meta File path to metadata (in Excel format). Default points to a metadata file for the Deerland Study.
#' @param sample_column The column name in the metadata file corresponding to sample IDs. Default is `"ID"`.
#' @param sample_nam_prefix A prefix to add to sample names. Default is `"S_"`.
#' @param rm_un Logical; whether to remove taxonomy entries starting with `"UN"`. Default is `TRUE`.
#' @param return_unstrat Logical; whether to return only unstratified pathways. Default is `TRUE`.
#' @param str_trun An integer specifying the maximum length of pathway names for truncation. Default is `40`.
#' @param add_MetaCyc_pathway_map Logical; whether to include MetaCyc pathway mapping. Default is `TRUE`.
#' @param add_CHOCOPhlAn_taxonomy Logical; whether to annotate with CHOCOPhlAn taxonomy. Default is `TRUE`.
#' @param final_tax_table_ranks A vector of column names defining the final structure of the taxonomy table. Default includes taxonomic and pathway features.
#' 
#' @return A `phyloseq` object with processed HUMAnN pathway data, cleaned sample names, taxonomy table, and metadata.
#' 
#' @details 
#' This function performs the following operations:
#' - Imports HUMAnN pathway data using `mia::importHUMAnN`.
#' - Converts the data into a `phyloseq` object for downstream microbial ecology analysis.
#' - Cleans sample names based on specified substitution patterns and regular expressions.
#' - Adds metadata to the `phyloseq` object for sample-level annotations.
#' - Processes and cleans the taxonomy table, replacing "unclassified" entries with `NA`, and optionally removing entries starting with "UN".
#' - Annotates pathways using MetaCyc and/or CHOCOPhlAn taxonomy data, with abbreviations for long pathway names and features.
#' 
#' @examples
#' \dontrun{
#' pathway_data <- "path/to/humann_pathway_data.tsv"
#' processed_data <- import_humann_pathway_mia(
#'   pathway = pathway_data,
#'   meta = "path/to/metadata.xlsx",
#'   sample_column = "SampleID"
#' )
#' }
#' 
#' @import microeco file2meco mia phyloseq microViz stringr dplyr readxl here
#' @export
#' 
#' 
import_humann_pathway_mia <- function(pathway = NULL,
                                      rank = "pathway",
                                      tax_clean_gsub = "[a-t]__",
                                      no_sel = c(NA, "<NA>","NA", "unclassified"),
                                      sub_pat_a = "_trimmed_HF_merged_Abundance-RELAB",
                                      sub_pat_b = "_subset50",
                                      tax_na_if = "unclassified",
                                      sample_name_clean_a = "[^_]+",
                                      sample_name_clean_b = "[A-Z]",
                                      meta = here::here("../../data/processed_data/metaphlan/Metadata_Deerland_Study.xlsx") %>% 
                                        read_excel(sheet = "Trial2"),
                                      sample_column = "ID",
                                      sample_nam_prefix = "S_",
                                      rm_un = TRUE,
                                      return_unstrat = TRUE,
                                      str_trun = 40,
                                      add_MetaCyc_pathway_map = TRUE,
                                      add_CHOCOPhlAn_taxonomy = TRUE,
                                      final_tax_table_ranks = c("Superclass1", "Superclass2", "feature", "Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")){
  
  
  suppressMessages(suppressWarnings({
    # Load required libraries
    require(microeco); require(file2meco) # For ecological and microbial ecology data analysis
    # require(microViz)   # An extension of phyloseq for speedy data manipulation
    
    pathway %>% 
      mia::importHUMAnN()  %>% 
      mia::convertToPhyloseq() %>% 
      # Clean sample names by applying various substitution patterns iteratively
      clean_phyloseq_sample_names(sub_pat = sub_pat_a) %>% 
      clean_phyloseq_sample_names(sub_pat = sub_pat_b) -> pw_mia_ps
    
    colnames(tax_table(pw_mia_ps)) = str_to_title(rank_names(pw_mia_ps))
    # taxa_sums(pw_mia_ps) %>%  sort(decreasing = T) %>%  tail()
    
    # Extract and clean sample names
    pw_mia_ps %>% 
      sample_names() %>% 
      str_extract(., sample_name_clean_a) %>%  # Extract meaningful parts of sample names
      str_remove_all(., sample_name_clean_b) -> sample_names(pw_mia_ps)
    
    # Add a prefix to sample names
    pw_mia_ps %>% 
      sample_names() %>% 
      paste0(sample_nam_prefix, .) -> sample_names(pw_mia_ps)
    
    # Add metadata to the phyloseq object
    pw_mia_ps %>% 
      physeq_add_metadata(metadata = meta, 
                          sample_column = sample_column) -> pw_mia_ps
    
    # Clean up the taxonomy table by removing unwanted patterns
    tax_table(pw_mia_ps) <- tax_table(pw_mia_ps) %>% 
      gsub(pattern = tax_clean_gsub, replacement = "") %>%  # Remove specific patterns
      # as.matrix() %>%  
      tax_table()
    
    # Replace "unclassified" entries with NA in the taxonomy table
    tax_table(pw_mia_ps) <- tax_table(pw_mia_ps) %>% 
      data.frame() %>% 
      mutate_all(na_if, tax_na_if) %>%  # Convert "unclassified" to NA
      as.matrix() %>%  
      tax_table()
    
    # Filter taxonomy table based on conditions (commented lines suggest alternative filters)
    pw_mia_ps %>%
      microViz::tax_mutate(feature = taxa_names(.)) -> pw_mia_ps
    
    # rank="feature"
    # pw_mia_ps %>% 
    #   speedyseq::filter_tax_table(get(rank) %!in% eval(as.character(no_sel))) -> pw_mia_ps
    # tax_mutate(Strain = NULL) %>%
    # microViz::tax_select(ranks_searched = rank, tax_list = eval(as.character(no_sel)),
    # deselect = TRUE) -> pw_me_ps_2              # Optional taxonomic selection step
    ## speedyseq::filter_tax_table(!eval(as.name((rank))) %in% no_sel) -> pw_me_ps# Exclude unwanted taxonomic ranks
    ## speedyseq::filter_tax_table(is.na("pathway")) %>%     # Alternative: filter NA pathways
    
    # pw_me_ps %>% 
    #   subset_taxa(is.na(Genus)) -> pw_me_ps
    
    if(return_unstrat)
    {  
      rank="Genus"
      no_sel=c(NA,"NA", "<NA>", NA_character_)
      
      pw_mia_ps %>% 
        speedyseq::filter_tax_table(get(rank) %in% eval(as.character(no_sel))) %>% 
        subset_taxa(!(grepl("unclassified", feature))) -> pw_mia_ps
    }
    
    if(rm_un)
    {
      pw_mia_ps %>%
        subset_taxa(!(grepl("^UN", feature)))  -> pw_mia_ps
      
      pw_mia_ps %>%
        microViz::tax_mutate(feature_code = taxa_names(.)) -> pw_mia_ps
      # Return the cleaned and updated phyloseq object
    }
    
    if(isTRUE(rm_un) & isTRUE(return_unstrat)){
      taxa_names(pw_mia_ps) <- taxa_names(pw_mia_ps) %>% 
        stringi::stri_extract(., regex = '[^:]*') 
    }
    
    if(add_MetaCyc_pathway_map)
    {
      data(MetaCyc_pathway_map)  
      
      pw_mia_ps %>%
        microViz::tax_mutate(feature_code = taxa_names(.)) -> pw_mia_ps
      
      pw_mia_ps %>% 
        tax_table() %>% 
        as.data.frame() %>% 
        rownames_to_column("tmp") %>% 
        # select(feature) %>% 
        dplyr::left_join(.,
                         MetaCyc_pathway_map %>% 
                           rownames_to_column("tmp") %>% 
                           mutate_all(funs(str_replace_all(., "&&", "-"))) %>% 
                           mutate_all(funs(str_replace_all(., "[&$|;]", "-"))),
                         by = c("tmp" = "tmp") 
        ) %>%  select(one_of(c("tmp", colnames(MetaCyc_pathway_map), rank_names(pw_mia_ps)))) %>% 
        column_to_rownames('tmp') %>% 
        mutate(feature = 
                 stringr::str_replace_all(pathway, 
                                          c(biosynthesis="Bios.", 
                                            vitamin="Vit.",
                                            metabolism = "Met.",
                                            nucleotides = "Nucleot.",
                                            nucleosides = "Nucleos.",
                                            carbohydrate = "Carb.",
                                            glycolysis = "Glycol.",
                                            assimilation = "Assim.",
                                            degradation = "Degrad.",
                                            `amino acid` = "AA",
                                            superpathway = "SPWY.",
                                            fermentation = "Ferm.",
                                            nutrient = "Nutr.",
                                            photosynthesis = "Photosynth.",
                                            pathways = "PWY.",
                                            degradation = "Deg.",
                                            utilization = "Util.",
                                            structure = "Stru.",
                                            vitamin = "Vit.",
                                            `fatty Acid` = "F.Acid",
                                            inorganic = "Inorg.",
                                            aromatic = "Aroma.",
                                            compound = "Comp.",
                                            decarboxylation = "decarboxyl.",
                                            reduction = "Red.",
                                            transformations = "transf.",
                                            photosynthetic = "Photosynth."))) %>% 
        mutate(across(c("Superclass1", "Superclass2"), 
                      ~ stringr::str_replace_all(., c(Precursor = "Precurs.",
                                                      Biosynthesis="Bios.", 
                                                      Transfer = "Trans.",
                                                      Vitamin="Vit.",
                                                      Generation = "Gen.",
                                                      Metabolism = "Met.",
                                                      Nucleotides = "Nucleot.",
                                                      Nucleosides = "Nucleos.",
                                                      Carbohydrate = "Carb.",
                                                      Glycolysis = "Glycol.",
                                                      Assimilation = "Assim.",
                                                      Degradation = "Degrad.",
                                                      `Amino acid` = "AA",
                                                      Superpathway = "SPWY.",
                                                      Fermentation = "Ferm.",
                                                      Nutrient = "Nutr.",
                                                      Photosynthesis = "Photosynth.",
                                                      Pathways = "PWY.",
                                                      Degradation = "Deg.",
                                                      Utilization = "Util.",
                                                      Structure = "Stru.",
                                                      Vitamin = "Vit.",
                                                      `Fatty Acid` = "F.Acid",
                                                      Inorganic = "Inorg.",
                                                      Aromatic = "Aroma.",
                                                      Compound = "Comp.",
                                                      Decarboxylation = "decarboxyl.",
                                                      Reduction = "Red.",
                                                      Transformations = "transf.",
                                                      Photosynthetic = "Photosynth.")
                      ))) %>% 
        mutate(across(c("Superclass1", "Superclass2","pathway"), 
                      ~ stringr::str_trunc(., width = str_trun, side ="center"))) %>% 
        select(one_of(final_tax_table_ranks)) %>% 
        as.matrix() -> tax_table(pw_mia_ps)
    }
    
    if(add_CHOCOPhlAn_taxonomy){
      data(CHOCOPhlAn_taxonomy)
      
      CHOCOPhlAn_taxonomy %>%
        as.matrix() %>%
        gsub(pattern = tax_clean_gsub, replacement = "") %>%  # Remove specific patterns
        as.data.frame() -> CHOCOPhlAn_taxonomy
      
      pw_mia_ps %>% 
        tax_table() %>% 
        as.data.frame() %>% 
        rownames_to_column("tmp") %>% 
        dplyr::left_join(.,
                         CHOCOPhlAn_taxonomy,
                         # rownames_to_column("tmp"),
                         by = c("Genus" = "Genus")) %>% 
        select(one_of(c("tmp",final_tax_table_ranks))) %>% 
        column_to_rownames('tmp') %>% 
        as.matrix() -> tax_table(pw_mia_ps)
      
    }
    
    if(!isTRUE(add_CHOCOPhlAn_taxonomy) & !isTRUE(add_MetaCyc_pathway_map)){
      pw_mia_ps %>% 
        physeq_sel_tax_table(final_tax_table_ranks) -> pw_mia_ps
    }
    
  }))
  return(pw_mia_ps)
}

# stringr::str_replace("unknown", "un.") %>%
# stringr::str_trunc(str_trun,  side ="center")

# str_replace_all(MetaCyc_pathway_map$Superclass2, "[^a-zA-Z0-9]", "")
# str_replace_all(MetaCyc_pathway_map$Superclass2, "[&$]", "-")
# "[[:punct:]]"
```

Extract the metadata from our metaphlan phyloseq object:

```{r}
ps_up %>% 
  sample_data() %>% 
  data.frame() %>% 
  rownames_to_column('tmp') -> metaph_metadata
```

Import HumAnN partway data and combine with the metadata of our metaphlan object:

```{r}
pathway %>% 
  import_humann_pathway_mia(meta = metaph_metadata, 
                            sample_column = 'tmp',
                            rm_un = FALSE, # Remove UNMAPPED and UNINTEGRATED Pathways
                            return_unstrat = FALSE, # 
                            add_CHOCOPhlAn_taxonomy = TRUE, # add Metaphlan/humann taxonomic path 
                            add_MetaCyc_pathway_map = FALSE) -> pw_mia # add MetaCyc pathway hierarchy

pw_mia %>% 
  tax_table() %>% 
  head()

pw_mia %>% 
  tax_table() %>% 
  tail()
```
By default return Unstratified data:

```{r}
pathway %>% 
  import_humann_pathway_mia(meta = metaph_metadata, 
                            sample_column = 'tmp') -> pw_mia

pw_mia %>% 
  tax_table() %>% 
  head()
```


We can also import Gene family data e.g., KEGG:

```{r}
kegg %>% 
  import_humann_pathway_mia(meta = metaph_metadata, 
                            sample_column = 'tmp',
                            rm_un = TRUE, 
                            return_unstrat = FALSE,
                            add_CHOCOPhlAn_taxonomy = TRUE, 
                            add_MetaCyc_pathway_map = FALSE) -> kegg_mia # add_MetaCyc_pathway_map only for pathway level data

kegg_mia %>% 
  tax_table() %>% 
  head()
```
Let's focus on the unstratified data, then we can always go back to the the stratified data to check which taxa are contributing to particular functions:

```{r}
kegg %>% 
  import_humann_pathway_mia(meta = metaph_metadata, 
                            sample_column = 'tmp',
                            rm_un = TRUE, 
                            return_unstrat = TRUE,
                            add_CHOCOPhlAn_taxonomy = TRUE, 
                            add_MetaCyc_pathway_map = FALSE) -> kegg_mia

kegg_mia %>% 
  tax_table() %>% 
  head()
```

## Use previous function for alpha - beta - differentially abundant taxa/features (e.g., Pathway, gene) tests:

### Alpha:

#### PWY:

Let's focus on the unstratified data, then we can always go back to the the stratified data to check which taxa are contributing to particular functions:

```{r}
alpha_measures = c("observed", "diversity_shannon", "diversity_inverse_simpson", "evenness_pielou")

pw_mia %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  phyloseq_alphas() %>% 
  pivot_longer(cols = all_of(alpha_measures), values_to = "value", names_to = 'alphadiversiy', values_drop_na  = TRUE) %>%
  mutate(alphadiversiy = fct_relevel(alphadiversiy, alpha_measures)) %>% 
  phyloseq_explore_alpha(facet_formula = "alphadiversiy ~ Sample",
                         group_by_stats = c("alphadiversiy", "Sample"),
                         stat_formula = "value ~ Time",
                         padjust_method = "fdr") -> out_pway

ls(out_pway)
```

Metrics are not calculated on the taxa here but on the functional Pathways:

```{r}
out_pway$alpha_plot
```

#### KO:

We can do the same with Gene level information, e.g.: Kegg's Orthologues:

```{r}
kegg_mia %>%
  physeq_glom_rename(taxrank = "feature", 
                     rename_ASV = "feature") %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  phyloseq_alphas() %>% 
  pivot_longer(cols = all_of(alpha_measures), values_to = "value", names_to = 'alphadiversiy', values_drop_na  = TRUE) %>%
  mutate(alphadiversiy = fct_relevel(alphadiversiy, alpha_measures)) %>% 
  phyloseq_explore_alpha(facet_formula = "alphadiversiy ~ Sample",
                         group_by_stats = c("alphadiversiy", "Sample"),
                         stat_formula = "value ~ Time",
                         padjust_method = "fdr",
                         # stat_paired = FALSE,
                         ref_group_stat = "TP1") -> out_kegg

out_kegg$alpha_plot
```

### Beta:

#### PWY:

```{r, warning=FALSE, echo=FALSE}
pw_mia %>%
  # subset_taxa(Class != "UNCLASSIFIED") %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  phyloseq_compute_bdiv() -> beta


pw_mia %>% 
  # subset_taxa(Class != "UNCLASSIFIED") %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  microViz::dist_calc(., dist = "robust.aitchison") %>% 
  microViz::dist_get() %>% 
  magrittr::divide_by(100) -> beta$rAitchison

beta$bray <- NULL
beta$sorensen <- NULL
```

```{r, warning=FALSE, echo=FALSE}
pd <- position_dodge(0.3)

pw_mia %>%
  # subset_taxa(Class != "UNCLASSIFIED") %>% 
  # transform_sample_counts(function(x) x/sum(x) * 1) %>% 
  phyloseq_explore_beta(ps_up = .,
                        beta = beta,
                        color_group = "Time",
                        shape_group = "Sample",
                        alpha = NULL,
                        col_pal = time_pal,
                        fill_pal = time_pal,
                        path_group = "interaction(Sample,Subject)",
                        facet_formula = "Sample ~ .",
                        axis1 = 1,
                        axis2 = 2,
                        seed = 123,
                        permanova_terms = c("Sample", "Time"),
                        strata = "none", 
                        metadata_dist_boxplot = c("Subject", "Time", "Sample")) -> out
```

```{r}
out$PCOA
```

```{r}
out$pcoas
```

```{r}
out$Sample$Saliva
```
```{r}
out$dist_box
```

#### KO:

```{r}
kegg_mia %>%
  physeq_glom_rename(taxrank = "feature", 
                     rename_ASV = "feature")  -> ps_tmp

ps_tmp %>%
  # subset_taxa(Class != "UNCLASSIFIED") %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  phyloseq_compute_bdiv() -> beta


ps_tmp %>% 
  # subset_taxa(Class != "UNCLASSIFIED") %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  microViz::dist_calc(., dist = "robust.aitchison") %>% 
  microViz::dist_get() %>% 
  magrittr::divide_by(100) -> beta$rAitchison

beta$bray <- NULL
beta$sorensen <- NULL

ps_tmp %>%
  # subset_taxa(Class != "UNCLASSIFIED") %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  phyloseq_explore_beta(ps_up = .,
                        beta = beta,
                        color_group = "Time",
                        shape_group = "Sample",
                        alpha = NULL,
                        col_pal = time_pal,
                        fill_pal = time_pal,
                        path_group = "interaction(Sample,Subject)",
                        facet_formula = "Sample ~ .",
                        axis1 = 1,
                        axis2 = 2,
                        seed = 123,
                        permanova_terms = c("Sample", "Time"),
                        strata = "none", 
                        metadata_dist_boxplot = c("Subject", "Time", "Sample")) -> out
```


```{r}
out$Sample
```
# Feature visualisation/ differential abundance tests:

Since some of the functions for differential abundance analyses and taxa/feature visualization expect to find Kingdom, Phylum, Class, Order, Family, Genus and Species as `rank_names` we can check the structure of the data and the name of the columns organizing that hierarchy:

```{r}
pw_mia %>% 
  tax_table() %>% 
  head()
```

Here 'Superclass1', 'Superclass2' and  'feature' are the interesting ones:

So we can use tax_mutate to organise (and copy) those data into Kingdom, Phylum, Class, Order, Family, Genus and Species columns.

```{r}
pw_mia %>% 
  tax_mutate(Kingdom = "MetaCyc", 
             Phylum = Superclass1,
             Class = Superclass1,
             Order = Superclass1,
             Family = Superclass2,
             Genus = Superclass2,
             Species = feature) %>% 
  physeq_sel_tax_table(c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) -> ps_tmp
```

Then the next steps look familiar!

```{r}
ps_tmp %>% 
  phyloseq_top_heatmap_barplot(facet_formula = "Sample_Type ~ Time" , group_var = "Sample_Time",
                               tax_levels = c("Phylum", "Genus", "Species"),
                               ntax = 4, ntax_species = 20, plot_heights = c(2.2, 1.5, 6),
                               boxplot_main_group = "Family",
                               rm_unclassified = FALSE,
                               facet_by = c("Sample_Type", "Time"),
                               group_by = c("Sample_Type", "Time"),
                               facet_heat = "~ Sample_Type + Time ") -> t

t$heat_all
```

```{r}
t$p
```

```{r}
t$nested_legend
```


```{r, message=TRUE}

ps_tmp %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = ., 
                          approach = c("run_lefse"),
                          comp_group = "Sample_Type",
                          glom = "Species", # agglomerate data at the species level
                          taxa_rank = "Species", # perform comparison at the genus level
                          lefse_lda_cutoff = 3, # filter results so only strong differences are displayed 
                          pvalue_cutoff = 0.0001, # filter results so only strong differences are displayed 
                          palette = sample_pal) -> lefse_TP1


lefse_TP1$mmlefse_p
```



```{r}

sessionInfo()
```
