---
title: " Study II - Taxonomic visualisation and statistical evaluation of metaphlan profiles"
author: "Florentin Constancias"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document: 
    toc: yes
    keep_md: yes
---

```{r setup, include=FALSE}
rm(list = ls())
gc()
# invisible(lapply(paste0('package:', names(sessionInfo()$otherPkgs)), detach, character.only=TRUE, unload=TRUE))
# options(java.parameters = "-Xmx80000m")
```

```{r setup2, include=FALSE}
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
##knitr::opts_chunk$set(cache=TRUE)
# knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(dev = "ragg_png") # https://ragg.r-lib.org/index.html
##knitr::opts_chunk$set(fig.width = 6) #https://r4ds.had.co.nz/graphics-for-communication.html#figure-sizing
knitr::opts_chunk$set(fig.asp = 0.618)
knitr::opts_chunk$set(fig.show = "hold")
knitr::opts_chunk$set(fig.show = "70%")
knitr::opts_chunk$set(fig.align = "center")
```

```{r setup3, include=FALSE}
library(tidyverse)
library(readxl)
library(readr)
library(speedyseq)
library(microViz)
```

```{r setup4, include=FALSE}
'%!in%' <- function(x,y)!('%in%'(x,y))

source_dir = "https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/"
# source_dir = "~/Documents/GitHub/DivComAnalyses/R/"

source(paste0(source_dir,"phyloseq_taxa_tests.R"))
source(paste0(source_dir,"phyloseq_normalisation.R"))
source(paste0(source_dir,"phyloseq_heatmap.R"))
source(paste0(source_dir,"phyloseq_varia.R"))
source(paste0(source_dir,"phyloseq_microeco.R"))
```

```{r}
load(here::here("../../data/processed_data/metaphlan/01_data.Rdata"))
```

```{r}
pd <- position_dodge(0.3)
```

# Explore proportion of the dominant taxa:


Source the function `phyloseq_top_heatmap_barplot`:

```{r}
source("https://raw.githubusercontent.com/fconstancias/KU_Caroline/refs/heads/main/code/functions/phyloseq_functions.R")
```

below the Arguments of the function phyloseq_top_heatmap_barplot() + default values:

```{r}
# parameters:

#' @param ps_up A phyloseq object containing microbiome data.
#' @param group_var Character string for the grouping variable (default: "Sample_Time").
#' @param tax_levels Character vector of taxonomic levels to include in plots (default: c("Phylum", "Family", "Genus", "Species")).
#' @param ntax Integer specifying the number of taxa to display at each level (default: 5).
#' @param ntax_species Integer specifying the number of species-level taxa to display (default: 14).
#' @param plot_heights Numeric vector for the heights of the plot panels (default: c(1.4, 1.5, 4)).
#' @param plot_x Character string for the x-axis label in bar plots (default: "Subject").
#' @param facet_by Character vector specifying variables to use for faceting in heatmaps (default: c("Sample_Type", "Time")).
#' @param group_by Character vector for grouping in heatmaps (default: c("Sample_Type", "Time")).
#' @param facet_heat Formula for faceting heatmap plots (default: "~ Sample_Type + Time").
#' @param facet_formula Formula for faceting bar plots (default: "Sample_Type ~ Time").
#' @param rm_unclassified Logical indicating if unclassified taxa should be removed (default: TRUE).
#' @param barplot_level Character string for taxonomic level in bar plots (default: "Species").
#' @param boxplot_main_group Character string for the main grouping variable in box plots (default: "Class").

```

Run the function:

```{r}
ps_up %>% 
  phyloseq_top_heatmap_barplot(facet_formula = "Sample_Type ~ Time" , 
                               ntax = 5, ntax_species = 10, plot_heights = c(1.4, 1.4, 4),
                               boxplot_main_group = "Family") -> out
```

Output is a list which containts different objects:

```{r}
ls(out)
```


```{r, fig.width = 8,  fig.asp = 1.2}
out$heat_all
```

```{r, fig.width = 9,  fig.asp = 0.5}
out$bar_plot + 
  ggpubr::rotate_x_text(60)
```

```{r}
out$p
```

```{r, fig.width = 8,  fig.asp = 1}
out$nested_legend
```

```{r, fig.width = 14,  fig.asp = 0.2}
out$ps_sub
```

```{r, fig.width = 14,  fig.asp = 0.2}

out$ps_sub + facet_null() +
  facet_grid(rows = vars(Sample_Type), 
             cols = vars(cluster_Dtp2, Subject )) + theme(legend.position = "none") 

```


# Statistical test to find differential abundant (i.e., proporition) taxa:


Function `phyloseq_diff_abundance()` allows to run different methods, here are the arguments:

```{r}
#' @title Differiential feature analyses wrapper for Microbiome analyses
#' @author Florentin Constancias
#' @description 
#' This script performs a comparative analysis of various approaches for microbiome studies, 
#' including metagenomics, metabarcoding, metabolomics, and transcriptomics. The script 
#' calculates the number of publications, environments analyzed, and the methods applied.
#' 
#' @param ps_tmp A phyloseq object. Defaults to saliva samples (`ps_up %>% subset_samples(Sample == "Saliva")`).
#' @param approach Character vector specifying the analysis approach. Options include:
#'   - "run_lefse": Linear discriminant analysis (LEfSe)
#'   - "run_ancom": Analysis of composition of microbiomes (ANCOM)
#'   - "ancombc2": ANCOMBC2 method for differential abundance
#'   - "maaslin3": Multivariate analysis using Maaslin3
#'   - "trans_diff_rf": Transformation-based Random Forest
#'   - "classifier_rf": Random Forest classifier
#' @param glom Taxonomic rank for agglomeration (e.g., "Species"). Set to `NULL` to skip agglomeration.
#' @param unclassified_name Name for unclassified taxa (default: "UNCLASSIFIED").
#' @param taxa_rank Taxonomic rank for differential abundance (default: "all").
#' @param density Density metric to normalize counts. Default is "Quant".
#' @param comp_group Grouping variable for comparison (e.g., "Time").
#' @param palette Color palette for plots.
#' @param pvalue_cutoff p-value cutoff for significance.
#' @param p_adjust Method for p-value adjustment (e.g., "BH" for Benjamini-Hochberg).
#' @param lefse_* Parameters specific to the LEfSe method.
#' @param maaslin3_* Parameters specific to Maaslin3 analysis.
#' @param ancombc2_* Parameters specific to ANCOMBC2 analysis.
#' @param linda_* Parameters specific to LINDA (if used in the future).
#' @param rf_* Parameters specific to Random Forest methods.
#' @param trans_diff_rf_MeanDecreaseGini_cutoff Gini index cutoff for feature importance in Random Forest.

```


## Test differences at Baseline between Saliva and Plaque microbiome:

```{r, message=TRUE}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse", "run_ancom", "trans_diff_rf", "maaslin3", "ancombc2", "linda"),
                          glom = "Species",
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          formula =  "~ Sample",
                          ancombc2_fix_formula = "Sample",
                          ancombc2_rand_formula = NULL,
                          ancombc2_group = "Sample",
                          linda_formula = NULL,
                          linda_comp_group = "Sample",,
                          palette = sample_pal,
                          unclassified_name = "UNCLASSIFIED") -> TP1
```


```{r, message=TRUE}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse"),
                          glom = "Species",
                          lefse_lda_cutoff = 4,
                          pvalue_cutoff = 0.001,
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          palette = sample_pal) -> lefse_TP1


lefse_TP1$mmlefse_df %>% 
  DT::datatable()
```
```{r, fig.width = 7,  fig.asp = 1}
lefse_TP1$mmlefse_p
```


## Exploring output : Lefse

Dataframe:

```{r}
TP1$mmlefse_df %>% 
  DT::datatable()
```
Plot:

```{r, fig.width = 18,  fig.asp = 1.6}
TP1$mmlefse_p 
```

```{r, message=TRUE}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse"),
                          glom = "Species",
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          palette = sample_pal,
                          unclassified_name = "UNCLASSIFIED") -> TP1
```

```{r, include=FALSE}
TP1$mmlefse -> tmp

TP1$mmlefse %>% 
  marker_table() %>% 
  data.frame() %>% 
  filter(ef_lda >= 4 & padj <= 0.0001) ->  marker_table(tmp) 

tmp %>% 
  plot_ef_bar() + scale_fill_manual(values = sample_pal)

# plot_ef_bar(tmp)
```

```{r}
# TP1$mmlefse_p + coord_cartesian(xlim = c(4, 20))
```

```{r}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse"),
                          glom = "Species",
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          lefse_multigrp_strat = TRUE,
                          lefse_strict = "2",
                          palette = sample_pal) -> TP1_lefse


TP1_lefse$mmlefse_df %>% 
  DT::datatable()
```


## Exploring output : 

```{r}
TP1$mmancom_p
```

## Exploring output : 

```{r}
TP1$trans_diff_rf$g1
```

## Exploring output : 

```{r}
TP1$maaslin3$maaslin3_lefse_plot
```

```{r}

```



```{r}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  taxa_list_boxplot(ps = ,
                    tax_vector = taxa_sel,
                    x =  "Sample",
                    color = "Sample",
                    palette = sample_pal,
                    log10 = TRUE,
                    log2 = FALSE,
                    taxa_rank = "Species",
                    unclassified_name = "UNCLASSIFIED") -> box_maaslin3

box_maaslin3
```

## Exploring output : 


```{r}
TP1$trans_diff_rf$res_diff %>% 
  rownames_to_column("feature") %>% 
  # filter(grepl(feature, "__s"))
  dplyr::filter(., grepl("s__",feature)) %>% 
  tidyr::separate(feature,
                  c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>% 
  mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>% 
  select(MeanDecreaseGini, Species, P.unadj, P.adj) -> tmp_res_rf
```


```{r}
plot_taxa_selection(ps_tmp = ps_up %>% 
                      subset_samples(Time == "TP1"),
                    diff_ab_out = TP1$trans_diff_rf$res_diff ,
                    diff_ab_filter = 'MeanDecreaseGini > 0.001 & P.adj <= 0.001',
                    full_path_tax_res = TRUE,
                    palette = sample_pal,
                    facet_formula = "Phylum ~ Sample_Type") -> test

test$bar_plot
```


```{r}
tmp_res_rf %>% 
  pull(Species) %>% unique() -> taxa_sel

ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  taxa_list_boxplot(ps = ,
                    tax_vector = taxa_sel,
                    x =  "Sample",
                    color = "Sample",
                    palette = sample_pal,
                    log10 = TRUE,
                    log2 = FALSE,
                    taxa_rank = "Species",
                    unclassified_name = "UNCLASSIFIED") -> box_rf

box_rf
```



```{r}
# out$rf$res_feature_imp %>% 
#   rownames_to_column("feature") %>% 
#   # filter(grepl(feature, "__s"))
#   dplyr::filter(., grepl("s__",feature)) %>% 
#   tidyr::separate(feature,
#                   c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>% 
#   mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>% 
#   
#   
#   
# dplyr::filter(feature, !grepl("RTB",TrackingPixel))
```
```{r}
out$rf

g1 <- out$plot_diff_bar(use_number = 1:100)
# use Genus level for parameter taxa_level, if you want to use all taxa, change to "all"
# nresam = 1 and boots = 1 represent no bootstrapping and use all samples directly
t1 <- trans_diff$new(dataset = mt, method = "rf", group = "Group", taxa_level = "all", filter_thres = 0.05, nresam = 0.75, boots = 999)

# plot the MeanDecreaseGini bar
# group_order is designed to sort the groups
g1 <- t1$plot_diff_bar(use_number = 1:100)
# plot the abundance using same taxa in g
g2 <- t1$plot_diff_abund(select_taxa = t1$plot_diff_bar_taxa, plot_type = "barerrorbar", add_sig = TRUE, errorbar_addpoint = TRUE, errorbar_color_black = TRUE)
# now the y axis in g1 and g2 is same, so we can merge them
# remove g1 legend; remove g2 y axis text and ticks
g1 <- g1 + theme(legend.position = "none")
g2 <- g2 + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.border = element_blank())
p <- g1 %>% aplot::insert_right(g2)
p

```

```{r}
out$rf$res_feature_imp %>% 
  filter(TP3.pval <= 0.05)
```


```{r}
out$maaslin3_lefse_plot + facet_null() + facet_grid(. ~ model)
```


```{r}
ps_up %>% 
  physeq_glom_rename(speedyseq = TRUE, taxrank = taxa_rank, rename_ASV = taxa_rank) %>% 
  phyloseq_maaslin3(formula = formula,
                    fixed_effects = "Time",strata_effects = "Subject") -> maaslin3# random_effects = , group_effects = )


```

```{r}
ps_up %>% 
  physeq_glom_rename(speedyseq = TRUE, taxrank = taxa_rank, rename_ASV = taxa_rank) %>% 
  phyloseq_maaslin3(formula = formula,
                    fixed_effects = "Time",random_effects  = "Subject") -> maaslin3_rd# random_effects = , group_effects = )


```

```{r}
########## ----- microeco::ancombc2 

ps_count  %>% 
  subset_taxa(Kingdom != unclassified_name) %>% 
  phyloseq_diff(method = "ancombc2", taxa_level = ifelse(taxa_rank == "all", "Species",taxa_rank),  group = comp_group,  alpha = pvalue_cutoff, #fix_formula = comp_group,
                p_adjust_method = p_adjust, plot_pal = palette) -> microeco_ancombc2

microbiomeMarker::marker_table(mmancombc) %>%
  data.frame() 
########## ----- microeco::linda 

ps_tmp %>% 
  subset_taxa(Kingdom != unclassified_name) %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  phyloseq_diff(method = "linda", linda_formula =  formula , taxa_level = ifelse(taxa_rank == "all", "Species",taxa_rank) ,group = NULL,  alpha = pvalue_cutoff, #fix_formula = comp_group,
                p_adjust_method = p_adjust, plot_pal = palette) -> microeco_linda



if (ancom_taxa_rank == "all")
{
  microbiomeMarker::marker_table(mmancom) %>%
    data.frame() %>% 
    filter(grepl("s__", feature)) %>% 
    tidyr::separate(feature,
                    c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>% 
    mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>% 
    select(Species, ef_CLR_F_statistic, W, enrich_group) %>% 
    mutate(ef_CLR_F_statistic = as.double(ef_CLR_F_statistic)) %>% 
    arrange(enrich_group, -ef_CLR_F_statistic) -> diff_sp_genomes
}else
{
  microbiomeMarker::marker_table(mmancom) %>%
    data.frame() %>% 
    pull(feature) -> diff_sp_genomes
}

########## ----- microeco::ancombc2

ps_count %>% 
  phyloseq_diff(method = "ancombc2", taxa_level = ancom_taxa_rank,  fix_formula = comp_group, alpha = 0.05) -> microeco_ancombc2

microeco_ancombc2$res_diff %>% 
  dplyr::filter(diff == "TRUE") %>%  # passed_ss == "TRUE"
  dplyr::filter(., !grepl("Intercept", Factors)) %>% 
  tidyr::separate(Taxa,
                  c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>% 
  mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>% 
  pull(Species) -> diff_sp_ancom_bc2

ps_tmp %>% 
  subset_taxa(Kingdom != unclassified_name) %>% # 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  prune_taxa(taxa = diff_sp_ancom_bc2, x = .) -> ps_sign

taxa_names(ps_sign) <- tax_table(ps_sign)[,"Species"]

lapply(
  as.list(taxa_names(ps_sign)),
  FUN = phyloseq_boxplot_abundance,
  ps = ps_tmp,
  x= comp_group, color = comp_group, level = "Species", line=NULL, violin = FALSE, show.points = TRUE, colors = palette) -> boxplots

names(boxplots) <- taxa_names(ps_sign)


########## ----- microbiomeMarker::run_lefse

ps_tmp %>% 
  physeq_sel_tax_table(c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>% 
  subset_taxa(Kingdom != unclassified_name) %>% 
  microbiomeMarker::run_lefse(group = comp_group,
                              taxa_rank = lefse_taxa_rank,
                              sample_min = 5,
                              strict = ,
                              multigrp_strat = , # one-against one (more strict) or in a one-against all setting, default FALSE.
                              lda_cutoff = 2) -> mmlefse

microbiomeMarker::plot_ef_bar(mmlefse) + scale_fill_manual(values = palette) -> mmlefse_p

microbiomeMarker::marker_table(mmlefse) %>%
  data.frame() -> mmlefse_df

ps_tmp %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  prune_taxa(taxa = diff_sp_genomes, x = .) -> ps_sign

ps_sign %>% 
  phyloseq_ampvis_heatmap(ntax = Inf,
                          transform = "identity",
                          tax_aggregate = "Species",
                          group_by = "SampleID",
                          tax_add  = "Family",
                          facet_by = c(comp_group)) -> p_diff_heat_sp


########## ----- 

ps_tmp %>% 
  subset_taxa(Kingdom != unclassified_name) %>% # 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  prune_taxa(taxa = diff_sp_genomes, x = .) -> ps_sign

taxa_names(ps_sign) <- tax_table(ps_sign)[,"Species"]

lapply(
  as.list(taxa_names(ps_sign)),
  FUN = phyloseq_boxplot_abundance,
  ps = ps_tmp,
  x= comp_group, color = comp_group, level = "Species", line=NULL, violin = FALSE, show.points = TRUE, colors = palette) -> boxplots

names(boxplots) <- taxa_names(ps_sign)



########## ----- 

```

```{r}

physeq_tax_only %>%
  subset_taxa(Order != "unassigned") %>% 
  microbiomeMarker::run_ancom(group = "group", 
                              taxa_rank = ancom_taxa_rank, 
                              confounders = ancom_confounders) -> ancom_age_all

plot_ef_bar(ancom_age_all) -> ancom_p_age_all

ancom_p_age_all + scale_fill_manual(values = group_pal) -> ancom_p_age_all

# ancom_age_all_sp <- ancom_age_all

# ancom_p_age_all$data %>%
#     # data.frame() %>% 
#     filter(grepl("s__", feature)) %>% 
#     tidyr::separate(feature,
#                     c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>% 
#     mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>% 
#     mutate(feature = Species) -> ancom_p_age_all$data
#   

marker_table(ancom_age_all) %>%
  data.frame() %>% 
  filter(grepl("s__", feature)) %>% 
  tidyr::separate(feature,
                  c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>% 
  mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>% 
  select(Species, ef_CLR_diff_mean, W, enrich_group) %>% 
  right_join(tax_plus_genome_info,
             by = c("Species" = "Species")) %>% 
  mutate(ef_CLR_diff_mean = as.double(ef_CLR_diff_mean)) %>% 
  arrange(enrich_group, -ef_CLR_diff_mean) -> diff_sp_genomes

physeq_tax_only %>% 
  transform_sample_counts(function(x) x/sum(x) * 100) %>% 
  prune_taxa(taxa = diff_sp_genomes %>%  
               dplyr::filter(!is.na(ef_CLR_diff_mean)) %>% 
               pull(Species), x = .) -> ps_sign

ps_sign %>% 
  phyloseq_ampvis_heatmap(ntax = Inf,
                          transform = "identity",
                          tax_aggregate = "Species",
                          group_by = "SampleID",
                          tax_add  = "Family",
                          facet_by = c("group")) -> p_diff_heat_sp

lapply(
  as.list(taxa_names(ps_sign)),
  FUN = phyloseq_boxplot_abundance,
  ps = ps_sign,
  x= "group", color = "group", level = "Species", line=NULL, violin = FALSE, show.points = TRUE, colors = group_pal) -> boxplots

names(boxplots) <- taxa_names(ps_sign)

########## PPFOR VOR   ####################

phyloseq %>%
  physeq_sel_tax_table(c("K00169", "K00179")) %>%
  # tax_glom(tax_meta) %>%
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  speedyseq::psmelt() -> ps_melted

ps_melted %>%
  pivot_longer("K00169":"K00179", names_to = "KeggID", values_to = "kegg_pres_abs") %>%
  select(OTU, Sample, group, sex, Abundance, KeggID, kegg_pres_abs, PAA_ng_ml, PAG_ng_ml) %>%
  group_by(Sample, group,sex, KeggID, kegg_pres_abs) %>%
  summarise(proportion = sum(Abundance)) %>%
  filter(kegg_pres_abs != "no-hit") -> KO_prop

# gphic = subset_taxa(physeq1,  eval(as.name(level_tax)) == king_list)

# KO_prop %>%
#   microViz::comp_barplot(
#     tax_transform_for_plot = "identity",
#     label = NULL,   tax_level =  tax_meta,  merge_other = FALSE,
#     # tax_level = "Phylum: Family", n_taxa = length(hierarchicalPal),
#     tax_order = "name",
#     sample_order = "default") +
#   ylab("Proportion - %") + theme_light() + facet_grid(. ~ group, scales = "free", space = "free_x", drop = TRUE) -> bar_plot

# bar_plot

KO_prop %>%
  ggpubr::ggboxplot(y = "proportion", color = "group",
                    x = "group", add = "jitter") +
  facet_grid(. ~ KeggID, scales = "free", space = "free_x", drop = TRUE) +
  ggpubr::stat_compare_means(aes(label = paste0("p = ", after_stat(p.format)))) +
  ylab("Proportion of KO harbouring taxa - %") +
  scale_color_manual(values = group_pal) +
  scale_fill_manual(values = group_pal) -> box_plot

KO_prop %>%
  ggpubr::ggboxplot(y = "proportion", color = "group", shape = "sex",
                    x = "group") +
  facet_grid(. ~ KeggID, scales = "free", space = "free_x", drop = TRUE) +
  ggpubr::stat_compare_means(aes(label = paste0("p = ", after_stat(p.format)))) +
  ylab("Proportion of KO harbouring taxa - %") +
  scale_color_manual(values = group_pal) +
  scale_fill_manual(values = group_pal) -> box_plot_sex

# box_plot

KO_prop %>%
  ggpubr::compare_means(proportion ~ group,
                        group.by = "KeggID",
                        data = .) -> stats

KO_prop %>%
  ggpubr::compare_means(proportion ~ sex,
                        group.by = c("KeggID","group"),
                        data = .) -> stats_sex


KO_prop %>%
  left_join(ps_melted %>%
              select(Sample, PAA_ng_ml, PAG_ng_ml)) %>%
  pivot_longer("PAA_ng_ml":"PAG_ng_ml", names_to = "meta", values_to = "meta_concentration") %>%
  ggpubr::ggscatter(., x = "proportion",
                    y = "meta_concentration", #color = "group", #facet.by = "name",
                    size = 2) +#, #label = "donors", repel = TRUE,
  # font.label = c(8, "plain"),
  # color = "sex", shape = "low_high_group", # Points color, shape and size
  # add = "reg.line",  # Add regressin line
  # add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
  # conf.int = TRUE) + #, # Add confidence interval
  # cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
  # cor.coeff.args = list(method = "pearson",
  # label.x = 25, label.y= 0,
  # label.sep = "   ")) +
  ylab("concentration ng/mL") +
  facet_grid(meta ~ KeggID, scales = "free") +
  stat_regline_equation(label.y.npc = "center") +
  stat_cor(method = "pearson", cor.coef.name = "R") +
  stat_cor(method = "spearman", cor.coef.name = "rho", label.y.npc = "bottom") +
  geom_smooth(method = "lm", se = TRUE, colour  = "black", fill= "red") +
  xlab(paste0("Proportion of  harboring taxa - %")) -> corr

KO_prop %>%
  left_join(ps_melted %>%
              select(Sample, PAA_ng_ml, PAG_ng_ml)) %>%
  pivot_longer("PAA_ng_ml":"PAG_ng_ml", names_to = "meta", values_to = "meta_concentration") %>%
  ggpubr::ggscatter(., x = "proportion",
                    y = "meta_concentration", #color = "group", #facet.by = "name",
                    size = 2) +#, #label = "donors", repel = TRUE,
  # font.label = c(8, "plain"),
  # color = "sex", shape = "low_high_group", # Points color, shape and size
  # add = "reg.line",  # Add regressin line
  # add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
  # conf.int = TRUE) + #, # Add confidence interval
  # cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
  # cor.coeff.args = list(method = "pearson",
  # label.x = 25, label.y= 0,
  # label.sep = "   ")) +
  ylab("concentration ng/mL") +
  facet_grid(meta ~ KeggID, scales = "free") +
  # stat_regline_equation(label.y.npc = "center") +
  # stat_cor(method = "pearson", cor.coef.name = "R") +
  # stat_cor(method = "spearman", cor.coef.name = "rho", label.y.npc = "bottom") +
  geom_smooth(method = "lm", se = TRUE, colour  = "black", fill= "red") +
  xlab(paste0("Proportion of  harboring taxa - %")) -> corr_no_eq

KO_prop %>%
  left_join(ps_melted %>%
              select(Sample, PAA_ng_ml, PAG_ng_ml)) %>%
  pivot_longer("PAA_ng_ml":"PAG_ng_ml", names_to = "meta", values_to = "meta_concentration") %>%
  ggpubr::ggscatter(., x = "proportion",
                    y = "meta_concentration", #color = "group", #facet.by = "name",
                    size = 2, #label = "donors", repel = TRUE,
                    # font.label = c(8, "plain"),
                    color = "sex") +#, #shape = "low_high_group", # Points color, shape and size
  # add = "reg.line",  # Add regressin line
  # # add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
  # conf.int = TRUE, # Add confidence interval
  # cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
  # cor.coeff.args = list(method = "pearson",
  #                       label.x = 25, label.y= 0,
  #                       label.sep = "   ")) +
  ylab("concentration ng/mL") +
  facet_grid(meta ~ KeggID, scales = "free") +
  stat_regline_equation(label.y.npc = "center") +
  stat_cor(method = "pearson", cor.coef.name = "R") +
  stat_cor(method = "spearman", cor.coef.name = "rho", label.y.npc = "bottom") +
  geom_smooth(method = "lm", se = TRUE, colour  = "black", fill= "red") +
  xlab(paste0("Proportion of  harboring taxa - %")) -> corr_sex

########## return   ####################

return(out = list("alpha_plot_stat" = alpha_plot_stat,
                  "alpha_plot" = alpha_plot,
                  "alpha_stat" = alpha_stat,
                  "alpha_stat_sex" = alpha_stat_sex,
                  "alpha_div" = alpha_div,
                  "pcoas" = pcoas,
                  "expl_var" = expl_var,
                  "PCoA" = PCoA,
                  "permanova" = permanova,
                  "fam_fit" = fam_fit,
                  # "gen_fit" = gen_fit,
                  # "sp_fit" = sp_fit,
                  
                  
                  "heat_all"= all,
                  "ancom" = ancom_age_all,
                  "ancom_LDA" = ancom_p_age_all,
                  "ancom_df" = marker_table(ancom_age_all) %>% data.frame(),
                  "genom_table_info" = diff_sp_genomes %>%  
                    replace(is.na(.), ""),
                  "p_diff_heat_sp" = p_diff_heat_sp,
                  "boxplots" = boxplots,
                  "KO_prop" = KO_prop,
                  "KO_wide" = KO_prop %>%  pivot_wider(names_from = c(Sample, group), values_from = proportion),
                  "stats" = stats,
                  "stats_sex" = stats_sex,
                  "box_plot"= box_plot,
                  "box_plot_sex" = box_plot_sex,
                  # "bar_plot" = bar_plot,
                  "corr_no_eq" = corr_no_eq,
                  "corr" = corr,
                  "corr_sex" = corr_sex))


```



# Plaque:

```{r}
ps_up %>%
  tax_glom(taxrank = "Species") %>%
  physeq_sel_tax_table(c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>%
  subset_samples(Sample == "Plaque")  -> physeq_tmp
```

## Lefse:

```{r}
physeq_tmp %>%
  phyloseq_diff(method = "lefse", group = "Time", taxa_level = "all", p_adjust_method = "fdr", plot_pal = time_pal, add_sig_plot2  = TRUE) -> lefse_plaque



lefse_plaque$diff_bar <-   lefse_plaque$diff_bar + theme(legend.position = "none")
lefse_plaque$diff_abund  <- lefse_plaque$diff_abund  + scale_y_sqrt() + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.border = element_blank()) + ylab("Proportion (sqrt)")

lefse_plaque$diff_bar %>% aplot::insert_right(lefse_plaque$diff_abund, width = 0.5)
```
## ANCOMBC2

```{r}
physeq_tmp %>%
  phyloseq_density_normalize("Quant") %>%
  phyloseq_diff(method = "ancombc2", group = "Time", taxa_level = "Species", p_adjust_method = "fdr", plot_pal = time_pal, add_sig_plot2  = TRUE) -> ancombc2_plaque

ancombc2_plaque$diff_bar <-   ancombc2_plaque$diff_bar + theme(legend.position = "none")
ancombc2_plaque$diff_abund  <- ancombc2_plaque$diff_abund  + scale_y_sqrt() + theme(axis.text.y = element_blank(), axis.ticks.y = element_blank(), panel.border = element_blank()) + ylab("Proportion (sqrt)")

ancombc2_plaque$diff_bar %>% aplot::insert_right(lefse_plaque$diff_abund, width = 0.5)
```
```{r}

```


```{r}
# physeq_tmp %>%
#   run_lefse(.,taxa_rank = "Species", group = "Time", norm = "CPM",
#             kw_cutoff = 0.05, lda_cutoff = 2, multigrp_strat = TRUE, strict = "0") -> lef_out
#
#
# lef_out %>%
#   marker_table() %>%  data.frame() -> lefse_df
#
# plot_ef_bar(lef_out) +
#   scale_color_manual(name = "", values = time_pal,
#                      na.value = "black") +
#   scale_fill_manual(name = "", values = time_pal,
#                     na.value = "black")  -> lef_plot
#
# lef_plot
```

```{r}
# physeq_tmp %>%
#   run_lefse(.,taxa_rank = "Genus", group = "Time", subgroup = "TP1", norm = "CPM",
#             kw_cutoff = 0.05, lda_cutoff = 2, multigrp_strat = TRUE, strict = "0") -> lef_out
#
# lef_out %>%
#   marker_table() %>%  data.frame() -> lefse_df
#
# plot_ef_bar(lef_out) +
#   scale_color_manual(name = "", values = time_pal,
#                      na.value = "black") +
#   scale_fill_manual(name = "", values = time_pal,
#                     na.value = "black")  -> lef_plot
#
# lef_plot
```

## Ancom:

```{r}
physeq_tmp %>%
  subset_taxa(Order != "unassigned") %>%
  microbiomeMarker::run_ancom(group = "Time",
                              taxa_rank = "Species",
                              p_adjust = "fdr") -> ancom #,
# confounders = c("age", "sex")) -> ancom

plot_ef_bar(ancom) -> ancom_p

ancom_p + scale_fill_manual(values = time_pal) -> ancom_p

ancom_p
```


```{r}
# ancom_age_all_sp <- ancom_age_all

# ancom_p_age_all$data %>%
#     # data.frame() %>%
#     filter(grepl("s__", feature)) %>%
#     tidyr::separate(feature,
#                     c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>%
#     mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>%
#     mutate(feature = Species) -> ancom_p_age_all$data
#

marker_table(ancom) %>%
  data.frame() %>%
  # filter(grepl("s__", feature)) %>%
  # tidyr::separate(feature,
  #                 c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>%
  # mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>%
  select(feature, ef_CLR_F_statistic, W, enrich_group) %>%
  # right_join(tax_plus_genome_info,
  #            by = c("Species" = "Species")) %>%
  mutate(ef_CLR_F_statistic = as.double(ef_CLR_F_statistic)) %>%
  arrange(enrich_group, -ef_CLR_F_statistic) -> diff_sp_genomes

physeq_tmp %>%
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  microViz::tax_select(ranks_searched = "Species", diff_sp_genomes %>%
                         dplyr::filter(!is.na(ef_CLR_F_statistic)) %>%
                         pull(feature)) -> ps_sign

ps_sign %>%
  phyloseq_ampvis_heatmap(ntax = Inf,
                          transform = "identity",
                          tax_aggregate = "Species",
                          group_by = "SampleID",
                          tax_add  = "Family",
                          facet_by = c("Time")) -> p_diff_heat_sp

# lapply(
#   as.list(taxa_names(ps_sign)),
#   FUN = phyloseq_boxplot_abundance,
#   ps = ps_sign,
#   x= "Time", color = "Time", level = "Species", line=NULL, violin = FALSE, show.points = TRUE, colors = time_pal) -> boxplots
#
# names(boxplots) <- taxa_names(ps_sign)

ps_sign %>%
  microViz::comp_barplot(
    tax_transform_for_plot = "identity",
    label = NULL,   tax_level =  "Species",  merge_other = FALSE,
    # tax_level = "Phylum: Family", n_taxa = length(hierarchicalPal),
    tax_order = sum, # "name",
    sample_order = "bray") + #"default"
  ylab("Proportion - %") + theme_light() + facet_grid(. ~ Time, scales = "free", space = "free_x", drop = TRUE) -> bar_plot

bar_plot

```

## Maaslin2:

```{r}
# physeq_tmp %>%
#   phyloseq_Maaslin2(fixed_effects = "Time",
#                     reference = "TP1")
```

# Saliva:

```{r}
ps_up %>%
  physeq_sel_tax_table(c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")) %>%
  subset_samples(Sample == "Saliva") -> physeq_tmp
```

## Lefse:


```{r}
physeq_tmp %>%
  run_lefse(.,taxa_rank = "Species", group = "Time", norm = "CPM",
            kw_cutoff = 0.05, lda_cutoff = 2, multigrp_strat = TRUE, strict = "0") -> lef_out

lef_out %>%
  marker_table() %>%  data.frame() -> lefse_df

plot_ef_bar(lef_out) +
  scale_color_manual(name = "", values = time_pal,
                     na.value = "black") +
  scale_fill_manual(name = "", values = time_pal,
                    na.value = "black")  -> lef_plot

lef_plot
```
```{r}
physeq_tmp %>%
  run_lefse(.,taxa_rank = "Genus", group = "Time", subgroup = "TP1", norm = "CPM",
            kw_cutoff = 0.05, lda_cutoff = 2, multigrp_strat = TRUE, strict = "0") -> lef_out

lef_out %>%
  marker_table() %>%  data.frame() -> lefse_df

plot_ef_bar(lef_out) +
  scale_color_manual(name = "", values = time_pal,
                     na.value = "black") +
  scale_fill_manual(name = "", values = time_pal,
                    na.value = "black")  -> lef_plot

lef_plot
```

## Ancom:

```{r}
physeq_tmp %>%
  subset_taxa(Order != "unassigned") %>%
  microbiomeMarker::run_ancom(group = "Time",
                              taxa_rank = "Species",
                              p_adjust = "fdr") -> ancom #,
# confounders = c("age", "sex")) -> ancom

plot_ef_bar(ancom) -> ancom_p

ancom_p + scale_fill_manual(values = time_pal) -> ancom_p

ancom_p
```


```{r}
# ancom_age_all_sp <- ancom_age_all

# ancom_p_age_all$data %>%
#     # data.frame() %>%
#     filter(grepl("s__", feature)) %>%
#     tidyr::separate(feature,
#                     c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>%
#     mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>%
#     mutate(feature = Species) -> ancom_p_age_all$data
#

marker_table(ancom) %>%
  data.frame() %>%
  # filter(grepl("s__", feature)) %>%
  # tidyr::separate(feature,
  #                 c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"),sep = "\\|") %>%
  # mutate(across(everything(), gsub, pattern = "[a-s]__", replacement = "")) %>%
  select(feature, ef_CLR_F_statistic, W, enrich_group) %>%
  # right_join(tax_plus_genome_info,
  #            by = c("Species" = "Species")) %>%
  mutate(ef_CLR_F_statistic = as.double(ef_CLR_F_statistic)) %>%
  arrange(enrich_group, -ef_CLR_F_statistic) -> diff_sp_genomes

physeq_tmp %>%
  transform_sample_counts(function(x) x/sum(x) * 100) %>%
  microViz::tax_select(ranks_searched = "Species", diff_sp_genomes %>%
                         dplyr::filter(!is.na(ef_CLR_F_statistic)) %>%
                         pull(feature)) -> ps_sign

ps_sign %>%
  phyloseq_ampvis_heatmap(ntax = Inf,
                          transform = "identity",
                          tax_aggregate = "Species",
                          group_by = "SampleID",
                          tax_add  = "Family",
                          facet_by = c("Time")) -> p_diff_heat_sp

# lapply(
#   as.list(taxa_names(ps_sign)),
#   FUN = phyloseq_boxplot_abundance,
#   ps = ps_sign,
#   x= "Time", color = "Time", level = "Species", line=NULL, violin = FALSE, show.points = TRUE, colors = time_pal) -> boxplots
#
# names(boxplots) <- taxa_names(ps_sign)

ps_sign %>%
  microViz::comp_barplot(
    tax_transform_for_plot = "identity",
    label = NULL,   tax_level =  "Species",  merge_other = FALSE,
    # tax_level = "Phylum: Family", n_taxa = length(hierarchicalPal),
    tax_order = sum, # "name",
    sample_order = "bray") + #"default"
  ylab("Proportion - %") + theme_light() + facet_grid(. ~ Time, scales = "free", space = "free_x", drop = TRUE) -> bar_plot

bar_plot

```


```{r}
# save(period_pal,
#      sample_pal,
#      sex_pal,
#      sub_pal,
#      time_pal,
#      ps_up,
#      p1, p1_leg, p2, p2_leg, p3, p3_leg,
#      file = here::here("data/04_data.Rdata"))
```


```{r}
sessionInfo()
```
