---
title: " Study II - Taxonomic visualisation and statistical evaluation of metaphlan profiles"
author: "Florentin Constancias"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document: 
    toc: yes
    keep_md: yes
---

```{r setup, include=FALSE}
rm(list = ls())
gc()
# invisible(lapply(paste0('package:', names(sessionInfo()$otherPkgs)), detach, character.only=TRUE, unload=TRUE))
# options(java.parameters = "-Xmx80000m")
```

```{r setup2, include=FALSE}
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
##knitr::opts_chunk$set(cache=TRUE)
# knitr::opts_chunk$set(echo = TRUE)
# knitr::opts_chunk$set(dev = "ragg_png") # https://ragg.r-lib.org/index.html
##knitr::opts_chunk$set(fig.width = 6) #https://r4ds.had.co.nz/graphics-for-communication.html#figure-sizing
knitr::opts_chunk$set(fig.asp = 0.618)
knitr::opts_chunk$set(fig.show = "hold")
knitr::opts_chunk$set(fig.show = "70%")
knitr::opts_chunk$set(fig.align = "center")
```

```{r setup3, include=FALSE}
library(tidyverse)
library(readxl)
library(readr)
library(speedyseq)
library(microViz)
```

```{r setup4, include=FALSE}


source_dir = "https://raw.githubusercontent.com/fconstancias/DivComAnalyses/master/R/"
# source_dir = "~/Documents/GitHub/DivComAnalyses/R/"

source(paste0(source_dir,"phyloseq_taxa_tests.R"))
source(paste0(source_dir,"phyloseq_normalisation.R"))
source(paste0(source_dir,"phyloseq_heatmap.R"))
source(paste0(source_dir,"phyloseq_varia.R"))
source(paste0(source_dir,"phyloseq_microeco.R"))
```

```{r}
load(here::here("../../data/processed_data/metaphlan/01_data.Rdata"))
```

```{r}
pd <- position_dodge(0.3)
```

# Explore proportion of the dominant taxa:


Source the function `phyloseq_top_heatmap_barplot`:

```{r}
source("https://raw.githubusercontent.com/fconstancias/KU_Caroline/refs/heads/main/code/functions/phyloseq_functions.R")
```

below the Arguments of the function phyloseq_top_heatmap_barplot() + default values:

```{r}
# parameters:

#' @param ps_up A phyloseq object containing microbiome data.
#' @param group_var Character string for the grouping variable (default: "Sample_Time").
#' @param tax_levels Character vector of taxonomic levels to include in plots (default: c("Phylum", "Family", "Genus", "Species")).
#' @param ntax Integer specifying the number of taxa to display at each level (default: 5).
#' @param ntax_species Integer specifying the number of species-level taxa to display (default: 14).
#' @param plot_heights Numeric vector for the heights of the plot panels (default: c(1.4, 1.5, 4)).
#' @param plot_x Character string for the x-axis label in bar plots (default: "Subject").
#' @param facet_by Character vector specifying variables to use for faceting in heatmaps (default: c("Sample_Type", "Time")).
#' @param group_by Character vector for grouping in heatmaps (default: c("Sample_Type", "Time")).
#' @param facet_heat Formula for faceting heatmap plots (default: "~ Sample_Type + Time").
#' @param facet_formula Formula for faceting bar plots (default: "Sample_Type ~ Time").
#' @param rm_unclassified Logical indicating if unclassified taxa should be removed (default: TRUE).
#' @param barplot_level Character string for taxonomic level in bar plots (default: "Species").
#' @param boxplot_main_group Character string for the main grouping variable in box plots (default: "Class").

```

Run the function:

```{r}
ps_up %>% 
  phyloseq_top_heatmap_barplot(group_var = "Sample",
                               facet_formula = "Sample_Type ~ Time" , 
                               ntax = 5, ntax_species = 10, plot_heights = c(1.4, 1.4, 4),
                               boxplot_main_group = "Family") -> out
```


```{r}
#' Modify Heatmap Text Annotations
#'
#' A function to modify a heatmap by adding text annotations for abundance values 
#' and adjusting the color scale with a pseudo-log transformation.
#'
#' @param heat A heatmap object (default: `out$heat$Genus`) assumed to be a ggplot object.
#' @param size Numeric. Text size for annotations (default: 2).
#' @param scale Numeric vector. Scale values for the color legend (default: `c(0, 0.01, 1, 10, 50, 75, 100)`).
#'
#' @return A ggplot object with updated text annotations and adjusted color scale.
#'
#' @examples
#' # Assuming `out$heat$Genus` is a ggplot object with a valid `Abundance` column:
#' modified_heatmap <- mod_text_heatmap(heat = out$heat$Genus, size = 3)
#' print(modified_heatmap)
#'
mod_text_heatmap <- function(heat = out$heat$Genus,
                             size = 2,
                             scale = c(0, 0.01, 1, 10, 50, 75, 100)) {
  require(tidyverse)  # Data manipulation and ggplot2
  require(gginnards)  # Editing ggplot objects
  require(ggpubr)
  
  # Remove existing text annotations
  heat %>%
    gginnards::delete_layers(., match_type = "GeomText") -> tmp
  
  # Add updated text annotations and adjust scale
  tmp +
    geom_text(data = tmp$data,
              aes(label = ifelse(is.na(Abundance), "", round(Abundance * 100, 2))),
              size = size, color = "black") +
    scale_fill_viridis_c(
      breaks = scale / 100,
      labels = scale,
      trans = scales::pseudo_log_trans(sigma = 0.001),
      na.value = 'transparent'
    ) +
    ggpubr::rotate_x_text(45) +  # Rotate x-axis text for better readability
    ylab(NULL) -> heat_up
  
  # Return the modified heatmap
  return(heat_up)
}

```

```{r}
out$heat$Species %>% 
  mod_text_heatmap() -> t

t
```

Output is a list which containts different objects:

```{r}
ls(out)
```


```{r, fig.width = 8,  fig.asp = 1.2}
out$heat_all
```

```{r, fig.width = 9,  fig.asp = 0.5}
out$bar_plot + 
  ggpubr::rotate_x_text(60)
```

```{r}
out$p
```

```{r, fig.width = 8,  fig.asp = 1}
out$nested_legend
```

```{r, fig.width = 14,  fig.asp = 0.2}
out$ps_sub
```

```{r, fig.width = 14,  fig.asp = 0.2}

out$ps_sub + facet_null() +
  facet_grid(rows = vars(Sample_Type), 
             cols = vars(cluster_Dtp2, Subject )) + theme(legend.position = "none") 

```


# Statistical test to find differential abundant (i.e., proporition) taxa:


Function `phyloseq_diff_abundance()` allows to run different methods, here are the arguments:

```{r}
#' @title Differiential feature analyses wrapper for Microbiome analyses
#' @author Florentin Constancias
#' @description 
#' This script performs a comparative analysis of various approaches for microbiome studies, 
#' including metagenomics, metabarcoding, metabolomics, and transcriptomics. The script 
#' calculates the number of publications, environments analyzed, and the methods applied.
#' 
#' @param ps_tmp A phyloseq object. Defaults to saliva samples (`ps_up %>% subset_samples(Sample == "Saliva")`).
#' @param approach Character vector specifying the analysis approach. Options include:
#'   - "run_lefse": Linear discriminant analysis (LEfSe)
#'   - "run_ancom": Analysis of composition of microbiomes (ANCOM)
#'   - "ancombc2": ANCOMBC2 method for differential abundance
#'   - "maaslin3": Multivariate analysis using Maaslin3
#'   - "trans_diff_rf": Transformation-based Random Forest
#'   - "classifier_rf": Random Forest classifier
#' @param glom Taxonomic rank for agglomeration (e.g., "Species"). Set to `NULL` to skip agglomeration.
#' @param unclassified_name Name for unclassified taxa (default: "UNCLASSIFIED").
#' @param taxa_rank Taxonomic rank for differential abundance (default: "all").
#' @param density Density metric to normalize counts. Default is "Quant".
#' @param comp_group Grouping variable for comparison (e.g., "Time").
#' @param palette Color palette for plots.
#' @param pvalue_cutoff p-value cutoff for significance.
#' @param p_adjust Method for p-value adjustment (e.g., "BH" for Benjamini-Hochberg).
#' @param lefse_* Parameters specific to the LEfSe method.
#' @param maaslin3_* Parameters specific to Maaslin3 analysis.
#' @param ancombc2_* Parameters specific to ANCOMBC2 analysis.
#' @param linda_* Parameters specific to LINDA (if used in the future).
#' @param rf_* Parameters specific to Random Forest methods.
#' @param trans_diff_rf_MeanDecreaseGini_cutoff Gini index cutoff for feature importance in Random Forest.

```


## Test differences at Baseline between Saliva and Plaque microbiome:

```{r, message=TRUE}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse", "run_ancom", "trans_diff_rf", "maaslin3", "ancombc2", "linda"),
                          glom = "Species",
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          formula =  "~ Sample",
                          ancombc2_fix_formula = "Sample",
                          ancombc2_rand_formula = NULL,
                          ancombc2_group = "Sample",
                          linda_formula = NULL,
                          linda_comp_group = "Sample",,
                          palette = sample_pal,
                          unclassified_name = "UNCLASSIFIED") -> TP1
```


```{r, message=TRUE}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse"),
                          glom = "Species",
                          lefse_lda_cutoff = 4,
                          pvalue_cutoff = 0.001,
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          palette = sample_pal) -> lefse_TP1


lefse_TP1$mmlefse_df %>% 
  DT::datatable()
```
```{r, fig.width = 7,  fig.asp = 1}
lefse_TP1$mmlefse_p
```


## Exploring output : Lefse

Dataframe:

```{r}
TP1$mmlefse_df %>% 
  DT::datatable()
```
Plot:

```{r, fig.width = 18,  fig.asp = 1.6}
TP1$mmlefse_p 
```

```{r, message=TRUE}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse"),
                          glom = "Species",
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          palette = sample_pal,
                          unclassified_name = "UNCLASSIFIED") -> TP1
```

```{r, include=FALSE}
TP1$mmlefse -> tmp

TP1$mmlefse %>% 
  marker_table() %>% 
  data.frame() %>% 
  filter(ef_lda >= 4 & padj <= 0.0001) ->  marker_table(tmp) 

tmp %>% 
  plot_ef_bar() + scale_fill_manual(values = sample_pal)

# plot_ef_bar(tmp)
```

```{r}
# TP1$mmlefse_p + coord_cartesian(xlim = c(4, 20))
```

```{r}
ps_up %>% 
  subset_samples(Time == "TP1") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          approach = c("run_lefse"),
                          glom = "Species",
                          taxa_rank = "all", #"OTU"
                          comp_group = "Sample",
                          lefse_multigrp_strat = TRUE,
                          lefse_strict = "2",
                          palette = sample_pal) -> TP1_lefse


TP1_lefse$mmlefse_df %>% 
  DT::datatable()
```


## Exploring output : 

```{r}
TP1$mmancom_p
```

## Exploring output : 

```{r}
TP1$trans_diff_rf$g1
```

## Exploring output : 

```{r}
TP1$maaslin3$maaslin3_lefse_plot
```

## Test differences between time points in Saliva samples:

```{r, message=TRUE}
ps_up %>% 
  subset_samples(Sample == "Saliva") %>% 
  phyloseq_diff_abundance(ps_tmp = .,
                          lefse_multigrp_strat = TRUE,
                          approach = c("run_lefse"),
                          glom = "Species",
                          taxa_rank = "all", 
                          comp_group = "Time",
                          palette = time_pal) -> lefse_saliva


lefse_saliva$mmlefse_p
```

```{r}
plot_taxa_selection(diff_ab_out = lefse_saliva$mmlefse_df, 
                    ps_tmp = ps_up %>% 
                      subset_samples(Sample == "Saliva"),
                    full_path_tax_res = TRUE, 
                    plot_x = "Time",
                    diff_ab_filter = "ef_lda > 0 & padj <= 0.05", 
                    taxa_sel_col = "feature",
                    full_path_sel = "s__",
                    palette = time_pal,
                    full_path_row_to_col = "tmp",
                    comp_group = "Time") -> saliva_plots

saliva_plots$heat
```

```{r}
saliva_plots$boxplot$Capnocytophaga_ochracea
```


```{r}
# save(period_pal,
#      sample_pal,
#      sex_pal,
#      sub_pal,
#      time_pal,
#      ps_up,
#      p1, p1_leg, p2, p2_leg, p3, p3_leg,
#      file = here::here("data/04_data.Rdata"))
```

***https://www.bioconductor.org/packages/release/bioc/vignettes/dar/inst/doc/article.html***


```{r}
sessionInfo()
```
